source/src/localisation/cs_TrackableParser.cpp

```diff
@@ -124,136 +223,371 @@ void TrackableParser::handleScannedDevice(scanned_device_t* device) {
+		retvalptr->summaries[i].crc     = _parsingFilters[i]->runtimedata.crc;
+		retvalptr->summaries[i].flags   = _parsingFilters[i]->metadata.flags;
+		retvalptr->summaries[i].version = _parsingFilters[i]->metadata.version;
+	}
+
+	result.returnCode = ERR_SUCCESS;
+}
+
+// -------------------------------------------------------------
+// ---------------------- Utility functions --------------------
+// -------------------------------------------------------------
+void TrackableParser::startProgress() {
+	filterModificationInProgress = true;
+	_masterVersion               = 0;
+}
+
+uint16_t TrackableParser::masterCrc() {
+	uint16_t masterCrc = 0;
+	uint16_t* prevCrc  = nullptr;
+
+	for (auto trackingFilter : _parsingFilters) {
+		if (trackingFilter == nullptr) {
+			break;
+		}
+
+		LOGTrackableParserDebug("filter crc: %x", trackingFilter->runtimedata.crc);
+
+		// appends/applies this filters' crc onto master crc
+		masterCrc =
+				crc16(reinterpret_cast<const uint8_t*>(&trackingFilter->runtimedata.crc),
+					  sizeof(trackingFilter->runtimedata.crc),
+					  prevCrc);
+
+		prevCrc = &masterCrc;  // first iteration will have prevCrc == nullptr, all subsequent ones point to masterCrc.
+	}
+
+	LOGTrackableParserDebug("master crc: %x", masterCrc);
+
+	return masterCrc;
+}
+
+bool TrackableParser::checkFilterSizeConsistency() {
+	bool consistencyChecksFailed = false;
+
+	size_t index = 0;
+	while (index < _parsingFiltersCount) {
+		auto trackingFilter = _parsingFilters[index];
+
+		if (trackingFilter == nullptr) {
+			// early return: last filter has been handled.
+			break;
+		}
+
+		if (trackingFilter->runtimedata.crc == 0) {
+			// filter changed since commit.
+
+			size_t trackingFilterSize = getTotalSize(*trackingFilter);
+			auto cuckoo               = CuckooFilter(trackingFilter->filterdata);
+
+			if (trackingFilterSize != sizeof(tracking_filter_t) + cuckoo.bufferSize()) {
+				// the cuckoo filter size parameters do not match the allocated space.
+				// likely cause by malformed data on the host side.
+				// This check can't be performed earlier: as long as not all chuncks are
+				// received the filterdata may be in invalid state.
+
+				LOGTrackableParserWarn("Deallocating filter %u because of malformed cuckoofilter size: %u != %u",
+						trackingFilter->runtimedata.filterId,
+						trackingFilterSize,
+						sizeof(tracking_filter_t) + cuckoo.bufferSize());
+
+				deallocateParsingFilter(trackingFilter->runtimedata.filterId);
+				consistencyChecksFailed = true;
+
+				// intentionally skipping index++, deallocate shrinks the array we're looping over.
+				continue;
+			}
+		}
+
+		index++;
+	}
+
+	return consistencyChecksFailed;
+}
+
+void TrackableParser::computeCrcs() {
+	for (auto trackingFilter : _parsingFilters) {
+		if (trackingFilter == nullptr) {
+			break;
+		}
+
+		if (trackingFilter->runtimedata.crc == 0) {
+			// filter changed since commit.
+			trackingFilter->runtimedata.crc =
```

Comment:
no CRC of the filter data?

Reply:
`runtimedata.totalSize` takes into account the filter data, is that what you mean?

Reply:
add comment that this takes filterdata into account